src/google-classroom-extractor/google_classroom_extractor/mapping/user_submission_activities.py

```diff
@@ -0,0 +1,154 @@
+        Content: Content associated with the activity
+        EntityStatus: The status of the record
+        LMSSectionIdentifier: A unique numeric identifier assigned to the section
+        SourceSystem: The system code or name providing the user data
+        SourceSystemIdentifier: A unique number or alphanumeric code assigned to a
+            user by the source system
+        LMSUserIdentifier: A unique numeric identifier assigned to the user
+        LMSUserActivityIdentifier: A unique numeric identifier assigned to the user activity
+        CreateDate: Date this record was created
+        LastModifiedDate: Date this record was last updated
+    """
+    assert isinstance(submissions_df, DataFrame)
+    assert "submissionHistory" in submissions_df.columns
+    assert "id" in submissions_df.columns
+    assert "courseId" in submissions_df.columns
+    assert "courseWorkId" in submissions_df.columns
+
+    # convert json-like submissionHistory string to list of dicts
+    submissions_df["submissionHistory"] = submissions_df["submissionHistory"].apply(lambda json_like: json.loads(json_like.replace("'", '"')))
+
+    submissions_df = submissions_df[["id", "courseId", "courseWorkId", "submissionHistory"]]
+
+    # explode submissionHistory lists into rows with other columns duplicated
+    history_df = submissions_df.explode("submissionHistory")
+
+    # expand submissionHistory dicts (stateHistory and gradeHistory) into their own columns
+    history_df = history_df["submissionHistory"].apply(Series).merge(history_df, left_index=True, right_index=True, how='outer')
+    history_df.drop(columns=["submissionHistory"], inplace=True)
+
+    # expand stateHistory
+    state_history_df = concat([history_df, history_df["stateHistory"].apply(Series)], axis=1)
+    state_history_df.dropna(subset=["stateHistory"], inplace=True)
+
+    # enrich state_history_df
+    state_history_df["SourceSystemIdentifier"] = "S-" + state_history_df[
+        ["courseId", "courseWorkId", "id", "stateTimestamp"]
+    ].agg("-".join, axis=1)
+
+    state_history_df = state_history_df[
+        [
+            "SourceSystemIdentifier",
+            "id",
+            "stateTimestamp",
+            "state",
+            "courseId",
+            "actorUserId",
+        ]
+    ]
+
+    state_history_df = state_history_df.rename(
+        columns={
+            "stateTimestamp": "ActivityDateTime",
+            "state": "ActivityStatus",
+            "courseId": "LMSSectionIdentifier",
+            "id": "AssignmentIdentifier",
+            "actorUserId": "LMSUserIdentifier",
+        }
+    )
+
+    state_history_df["LMSUserActivityIdentifier"] = state_history_df["SourceSystemIdentifier"]
+    state_history_df["ActivityType"] = "Submission State Change"
+    state_history_df["ActivityTimeInMinutes"] = ""
+    state_history_df["Content"] = ""
+    state_history_df["EntityStatus"] = ENTITY_STATUS_ACTIVE
```

Comment:
I have yet to figure out the purpose of this field. I keep meaning to ask Gabrielle again. Thus there is never a case in Schoology where we're setting a value other than `active`. And I see that here with the user submissions. Are there any cases so far, in Google Classroom, where you have a value other than `active`?

Reply:
It's always Active so far.